presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -387,6 +385,15 @@ private void spillInput(Page page)
         checkState(spillEnabled, "Spill not enabled, no revokable memory should be reserved");
 
         if (state == State.CONSUMING_INPUT) {
+            long indexSizeBeforeCompaction = index.getEstimatedSize().toBytes();
+            index.compact();
+            long indexSizeAfterCompaction = index.getEstimatedSize().toBytes();
+            if (indexSizeAfterCompaction < indexSizeBeforeCompaction * 0.8) {
+                finishMemoryRevoke = Optional.of(() -> {
+                });
+                return immediateFuture(null);
+            }
```

Comment:
can we wrap this code with methods, that way the code would be more clear and readable as it would be on higher abstraction level:

```
if (tryToCompactIndex()) {
    spill();
}
```